src/sinks/datadog/traces/apm_stats/flusher.rs

```diff
@@ -0,0 +1,173 @@
+        aggregator,
+    };
+
+    // flush on the same interval as the stats buckets
+    let mut interval =
+        tokio::time::interval(std::time::Duration::from_nanos(BUCKET_DURATION_NANOSECONDS));
+
+    debug!("Starting APM stats flushing thread.");
+
+    loop {
+        tokio::select! {
+
+        _ = interval.tick() => {
+            // flush the oldest bucket from the cache to Datadog
+            sender.flush_apm_stats(false).await;
+        },
+        signal = &mut tripwire =>  match signal {
+            // sink has signaled us that the process is shutting down
+            Ok(sink_shutdown_ack_sender) => {
+
+                debug!("APM stats flushing thread received exit condition. Flushing remaining stats before exiting.");
+                sender.flush_apm_stats(true).await;
+
+                // signal the sink (who tripped the tripwire), that we are done flushing
+                let _ = sink_shutdown_ack_sender.send(());
+                break;
+            }
+            Err(_) => error!(message = "Tokio Sender unexpectedly dropped.", internal_log_rate_limit = true),
+        }
+        }
+    }
+}
+
+struct ApmStatsSender {
+    client: HttpClient,
+    compression: Compression,
+    endpoint_configuration: DatadogTracesEndpointConfiguration,
+    aggregator: Arc<Mutex<Aggregator>>,
+}
+
+impl ApmStatsSender {
+    async fn flush_apm_stats(&self, force: bool) {
+        // explicit scope to minimize duration that the Aggregator is locked.
+        if let Some((payload, api_key)) = {
+            let mut aggregator = self.aggregator.lock().unwrap();
+            let client_stats_payloads = aggregator.flush(force);
+
+            if client_stats_payloads.is_empty() {
+                // no sense proceeding if no payloads to flush
+                None
+            } else {
+                let payload = StatsPayload {
+                    agent_hostname: aggregator.get_agent_hostname(),
+                    agent_env: aggregator.get_agent_env(),
+                    stats: client_stats_payloads,
+                    agent_version: aggregator.get_agent_version(),
+                    client_computed: false,
+                };
+
+                Some((payload, aggregator.get_api_key()))
+            }
+        } {
+            if let Err(e) = self.compress_and_send(payload, api_key).await {
+                // TODO emit an internal `Error` event here, probably
+                error!(
+                    message = format!("Error while encoding and sending APM stats payloads: {}", e),
+                    internal_log_rate_limit = true,
+                );
+            }
+        }
+    }
+
+    async fn compress_and_send(
+        &self,
+        payload: StatsPayload,
+        api_key: Arc<str>,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let (metadata, compressed_payload) = self.build_apm_stats_request_data(api_key, payload)?;
+
+        let request_metadata = RequestMetadata::new(0, 0, 0, 0, 0);
```

Comment:
Is there no `Default` impl for `RequestMetadata`? `new(0, 0, 0, 0, 0)` vs `default` (or `empty`) feels a little silly, heh.

Reply:
:joy: whoops, yeah that was an oversight of moving too fast. Will use `::default()` :+1: